File: cpp/opendnp3/src/opendnp3/objects/Group32.cpp
//
//  _   _         ______    _ _ _   _             _ _ _
// | \ | |       |  ____|  | (_) | (_)           | | | |
// |  \| | ___   | |__   __| |_| |_ _ _ __   __ _| | | |
// | . ` |/ _ \  |  __| / _` | | __| | '_ \ / _` | | | |
// | |\  | (_) | | |___| (_| | | |_| | | | | (_| |_|_|_|
// |_| \_|\___/  |______\__,_|_|\__|_|_| |_|\__, (_|_|_)
//                                           __/ |
//                                          |___/
// Copyright 2013 Automatak LLC
// 
// Automatak LLC (www.automatak.com) licenses this file
// to you under the the Apache License Version 2.0 (the "License"):
// 
// http://www.apache.org/licenses/LICENSE-2.0.html
//

#include "Group32.h"

#include "../PackingUnpacking.h"

namespace opendnp3 {

Group32Var1 Group32Var1::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var1 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int32LE::Read(buffer);
  buffer.Advance(4);
  return obj;
}

Group32Var2 Group32Var2::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var2 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int16LE::Read(buffer);
  buffer.Advance(2);
  return obj;
}

Group32Var3 Group32Var3::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var3 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int32LE::Read(buffer);
  buffer.Advance(4);
  obj.time48 = UInt48LE::Read(buffer);
  buffer.Advance(6);
  return obj;
}

Group32Var4 Group32Var4::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var4 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = Int16LE::Read(buffer);
  buffer.Advance(2);
  obj.time48 = UInt48LE::Read(buffer);
  buffer.Advance(6);
  return obj;
}

Group32Var5 Group32Var5::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var5 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  return obj;
}

Group32Var6 Group32Var6::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var6 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  return obj;
}

Group32Var7 Group32Var7::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var7 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  obj.time48 = UInt48LE::Read(buffer);
  buffer.Advance(6);
  return obj;
}

Group32Var8 Group32Var8::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group32Var8 obj;
  obj.flags = UInt8::Read(buffer);
  buffer.Advance(1);
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  obj.time48 = UInt48LE::Read(buffer);
  buffer.Advance(6);
  return obj;
}


}

File: generation/dnp3/src/main/scala/com/automatak/render/dnp3/objects/groups/Group32.scala
/**
 * Copyright 2013-2022 Step Function I/O, LLC
 *
 * Licensed to Green Energy Corp (www.greenenergycorp.com) and Step Function I/O
 * LLC (www.automatak.com) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership. Green Energy Corp and Step Function I/O LLC license
 * this file to you under the Apache License, Version 2.0 (the "License"); you
 * may not use this file except in compliance with the License. You may obtain
 * a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.automatak.render.dnp3.objects.groups

import com.automatak.render.dnp3.objects._

import FixedSizeField._
import com.automatak.render.dnp3.objects.VariationNames._
import com.automatak.render.dnp3.objects.generators.ConversionToAnalog

// analog input event
object Group32 extends ObjectGroup {
  def objects = List(Group32Var0, Group32Var1, Group32Var2, Group32Var3, Group32Var4, Group32Var5, Group32Var6, Group32Var7, Group32Var8)
  def group: Byte = 32
  def desc: String = "Analog Input Event"
  def isEventGroup: Boolean = true
}

object Group32Var0 extends AnyVariation(Group32, 0)
object Group32Var1 extends FixedSize(Group32, 1, bit32WithFlag)(flags, value32) with ConversionToAnalog
object Group32Var2 extends FixedSize(Group32, 2, bit16WithFlag)(flags, value16) with ConversionToAnalog
object Group32Var3 extends FixedSize(Group32, 3, bit32WithFlagTime)(flags, value32, time48) with ConversionToAnalog
object Group32Var4 extends FixedSize(Group32, 4, bit16WithFlagTime)(flags, value16, time48) with ConversionToAnalog
object Group32Var5 extends FixedSize(Group32, 5, singlePrecisionWithFlag)(flags, float32) with ConversionToAnalog
object Group32Var6 extends FixedSize(Group32, 6, doublePrecisionWithFlag)(flags, float64) with ConversionToAnalog
object Group32Var7 extends FixedSize(Group32, 7, singlePrecisionWithFlagTime)(flags, float32, time48) with ConversionToAnalog
object Group32Var8 extends FixedSize(Group32, 8, doublePrecisionWithFlagTime)(flags, float64, time48) with ConversionToAnalog
